### ISSUE TITLE
Improve sampling in auto-level

### ISSUE THREAD

**Issue by jbvioix** [vlachoudis/bCNC#1206]
Wegstr has implemented auto-level to their machines. As you can see on the video (https://www.youtube.com/watch?v=-QqWOIUeAV4 near 3:00) auto level seems to use an irregular pattern not a regular grid. Maybe inflexion points on the tracks can be uses ?

**Comment by MARIOBASZ** [comment]
@jbvioix , It seems that you are using a concept like the driller plugin to generate the autollevel points. I do not know if this results in more or less time of scanning the surface.

**Comment by Harvie** [comment]
@jbvioix video link does not work...

**Comment by carl1961** [comment]
Harvie, if the circle board has a circle edge.nc could or would autolevel use that as a reference?

**Comment by MARIOBASZ** [comment]
Using a spiral path for test points can help?

**Comment by Harvie** [comment]
@carl1961 yes. that is the plan, we just need to implement it :-)

**Comment by jbvioix** [comment]
The video is https://www.youtube.com/watch?v=Mxa6xQFCyXA near 3:10 you can see the movement of the drill is not on a regular grid.

**Comment by MARIOBASZ** [comment]
One strategy that I can think of to accelerate the sampling, is to discard the points that are far from the traces, if the time of intersection processing allows it

**Comment by MARIOBASZ** [comment]
@jbvioix , By adapting driller, you can sample the routes at predetermined distances. It would be necessary to discard the points corresponding to nearby traces, it could be a good strategy

**Comment by MARIOBASZ** [comment]
Maybe following this thread:
of something like pirograph, or a raster for laser:
From this:
![image](https://user-images.githubusercontent.com/11743403/60690112-fd19f600-9e9a-11e9-8f30-8d153c722ac8.png)
to this
![image](https://user-images.githubusercontent.com/11743403/60690136-33577580-9e9b-11e9-96d8-b34dbeeed7a9.png)

**Comment by jbvioix** [comment]
The video is https://www.youtube.com/watch?v=Mxa6xQFCyXA near 3:10 you can see the movement of the drill is not on a regular grid.

**Comment by MARIOBASZ** [comment]
@jbvioix , I have seen it.
Do you think that strategy is an advantage?
![image](https://user-images.githubusercontent.com/11743403/61086815-41147a00-a40a-11e9-9aa0-85a9dab12403.png)
try scan as driller plugin, combining anchor with intermediate points ?
(it is not a good representation)
![image](https://user-images.githubusercontent.com/11743403/61087372-cd736c80-a40b-11e9-8556-999b596ff171.png)
![image](https://user-images.githubusercontent.com/11743403/61087381-dbc18880-a40b-11e9-95db-6f583d05f710.png)

**Comment by MARIOBASZ** [comment]
I think a good strategy is the current grid, with the condition "only if there are traces at close range"

**Comment by Donnyp751** [comment]
How WEGSTR auto levels is a very poor way of doing it. Not good for circuit boards. I have one of their machines. It literally just probes every N distance along the path of the gcode for a z height, then uses that as an offset. It changes the Z height when it reaches where it probed. It's not a very smooth transition. Causes sudden Z changes making it not isolate through some areas if the surface is at all curved.

**Comment by Danivas13** [comment]
I think that the speed can be increase modify de Z safe. Generate a new z safe for probbing just a 0.3mm above the last probe point and then move to the next point.

**Comment by Harvie** [comment]
> Generate a new z safe for probbing just a 0.3mm above the last probe point and then move to the next point.

We are not using probe just for flat surfaces autoleveling. We use it to scan 3D shapes as well. In such case the height can vary rapidly between the points.

**Comment by Danivas13** [comment]
> > Generate a new z safe for probbing just a 0.3mm above the last probe point and then move to the next point.
> 
> We are not using probe just for flat surfaces autoleveling. We use it to scan 3D shapes as well. In such case the height can vary rapidly between the points.

Yes. but you can especify de max in the probe settings, the only thing that change es where start to count (origin Z-0 or the preview probe-Z.
